src/dscanner/analysis/auto_function.d

```diff
@@ -20,218 +17,147 @@ import std.algorithm.searching : any;
  * to infer attributes.
  */
-final class AutoFunctionChecker : BaseAnalyzer
+extern(C++) class AutoFunctionChecker(AST) : BaseAnalyzerDmd
 {
-
-private:
-
-	enum string KEY = "dscanner.suspicious.missing_return";
-	enum string MESSAGE = "Auto function without return statement, prefer an explicit void";
-
-	bool[] _returns;
-	size_t _mixinDepth;
-	string[] _literalWithReturn;
-
-public:
-
-	alias visit = BaseAnalyzer.visit;
-
+	alias visit = BaseAnalyzerDmd.visit;
 	mixin AnalyzerInfo!"auto_function_check";
 
 	///
-	this(string fileName, bool skipTests = false)
+	extern(D) this(string fileName, bool skipTests = false)
 	{
-		super(fileName, null, skipTests);
+		super(fileName, skipTests);
 	}
 
-	override void visit(const(FunctionDeclaration) decl)
+	override void visit(AST.FuncDeclaration d)
 	{
-		_returns.length += 1;
-		scope(exit) _returns.length -= 1;
-		_returns[$-1] = false;
-
-		const bool autoFun = decl.storageClasses
-			.any!(a => a.token.type == tok!"auto" || a.atAttribute !is null);
+		import dmd.astenums : STC, STMT;
 
-		decl.accept(this);
+		if (d.storage_class & STC.disable)
+			return;
 
-		if (decl.functionBody.specifiedFunctionBody && autoFun && !_returns[$-1])
-			addErrorMessage(decl.name.line, decl.name.column, KEY, MESSAGE);
-	}
-
-	override void visit(const(ReturnStatement) rst)
-	{
-		if (_returns.length)
-			_returns[$-1] = true;
-		rst.accept(this);
-	}
-
-	override void visit(const(AssertArguments) exp)
-	{
-		exp.accept(this);
-		if (_returns.length)
+		if (!(d.storage_class & STC.auto_) && !d.inferRetType)
 		{
-			const UnaryExpression u = cast(UnaryExpression) exp.assertion;
-			if (!u)
-				return;
-			const PrimaryExpression p = u.primaryExpression;
-			if (!p)
-				return;
-
-			immutable token = p.primary;
-			if (token.type == tok!"false")
-				_returns[$-1] = true;
-			else if (token.text == "0")
-				_returns[$-1] = true;
+			super.visitFuncBody(d);
+			return;
 		}
-	}
-
-	override void visit(const(MixinExpression) mix)
-	{
-		++_mixinDepth;
-		mix.accept(this);
-		--_mixinDepth;
-	}
 
-	override void visit(const(PrimaryExpression) exp)
-	{
-		exp.accept(this);
+		// arrow functions
+		if (auto rs = d.fbody.isReturnStatement())
+			return;
 
-		import std.algorithm.searching : canFind;
+		// // fbody is either return or compound statement
+		auto cs = d.fbody.isCompoundStatement();
 
-		if (_returns.length && _mixinDepth)
+		// look for asser(0) or assert(false)
+		foreach (s; *cs.statements)
 		{
-			if (findReturnInLiteral(exp.primary.text))
-			    _returns[$-1] = true;
-			else if (exp.identifierOrTemplateInstance &&
-				_literalWithReturn.canFind(exp.identifierOrTemplateInstance.identifier.text))
-					_returns[$-1] = true;
+			AST.ExpStatement es = s.isExpStatement();
+			AST.AssertExp ae = null;
+
+			if (es && es.exp)
+				ae = es.exp.isAssertExp();
+
+			if (ae)
+			{
+				auto ie = ae.e1.isIntegerExp();
+
+				if (ie && ie.getInteger() == 0)
+				{
+					super.visitFuncBody(d);
```

Comment:
I don't understand the logic here. If you encounter an `assert(0)` you start visiting the function body again?

Reply:
No, I am not visiting again. The idea is that if a function has an `assert(0)` it should be exempt from the check. However, it's children(potentially other functions) should be checked, that's why the `super.visit` call. Example scenario:
```
auto doStuff1() // shouldn't warn
{
    auto doStuff2(){} // should warn
    assert(0);
}
```

Reply:
But you have already started visiting the function body by iterating through `cs`. It seems to me that you are doing extra work. Why not just visit each individual statement of `cs`? 

A simpler logic would be to simply check each statement inside a function body and treat each case:

- if it's a func decl, just check it
- if it's a return statement, set foundReturnStatement
- if it's an assert set foundReturnStatement